lib/active_triples/relation.rb

```diff
@@ -327,33 +343,110 @@ def property
     end
 
     ##
-    # Gives the predicate used by the Relation. Values of this object are
-    # those that match the pattern `<rdf_subject> <predicate> [value] .`
+    # Adds values to the relation
     #
-    # @return [RDF::Term, nil] the predicate for this relation; nil if
-    #   no predicate can be found
+    # @param [Array<RDF::Resource>, RDF::Resource] values  an array of values
+    #   or a single value. If not an {RDF::Resource}, the values will be
+    #   coerced to an {RDF::Literal} or {RDF::Node} by {RDF::Statement}
     #
-    # @see #property
-    def predicate
-      return property if property.is_a?(RDF::Term)
-      property_config[:predicate] if is_property?
+    # @return [Relation] a relation containing the set values; i.e. `self`
+    #
+    # @raise [ActiveTriples::UndefinedPropertyError] if the property is not
+    #   already an {RDF::Term} and is not defined in `#property_config`
+    #
+    # @see http://www.rubydoc.info/github/ruby-rdf/rdf/RDF/Statement For
+    #   documentation on {RDF::Statement} and the handling of
+    #   non-{RDF::Resource} values.
+    def set(values)
+      raise UndefinedPropertyError.new(property, reflections) if predicate.nil?
+      values = values.to_a if values.is_a? Relation
+      values = [values].compact unless values.kind_of?(Array)
+
+      clear
+      values.each { |val| set_value(val) }
+
+      parent.persist! if parent.persistence_strategy.is_a? ParentStrategy
+      self
+    end
+
+    ##
+    # @overload subtract(enum)
+    #   Deletes objects in the enumerable from the relation
+    #   @param values [Enumerable] an enumerable of objects to delete
+    # @overload subtract(*values)
+    #   Deletes each argument value from the relation
+    #   @param *values [Array<Object>] the objects to delete
+    #
+    # @return [Relation] self
+    #
+    # @note This casts symbols to a literals, which gets us symmetric behavior
+    #   with `#set(:sym)`.
+    # @see #delete
+    def subtract(*values)
+      values = values.first if values.first.is_a? Enumerable
+      statements = values.map do |value|
+        value = RDF::Literal(value) if value.is_a? Symbol
+        [rdf_subject, predicate, value]
+      end
+
+      parent.delete(*statements)
+      self
+    end
+
+    ##
+    # Replaces the first argument with the second as a value within the
+    # relation.
+    #
+    # @param swap_out [Object] the value to delete
+    # @param swap_in  [Object] the replacement value
+    #
+    # @return [Relation] self
+    def swap(swap_out, swap_in)
+      self.<<(swap_in) if delete?(swap_out)
     end
 
     protected
 
+      ##
+      # @private
       def node_cache
         @node_cache ||= {}
       end
 
+      ##
+      # @private
+      def objects(&block)
+        solutions = parent.query(subject: rdf_subject, predicate: predicate)
+        solutions.extend(RDF::Enumerable) unless solutions.respond_to?(:each_object)
```

Comment:
This requires some explanation. We need to include this extension behavior because the `RDF::Queryable#query` [return values](https://github.com/ruby-rdf/rdf/blob/develop/lib/rdf/mixin/queryable.rb#L40-L41) are pretty loosy goosy. In some cases we can end up with a `Enumerator<RDF::Statement>` (usually an `Array`) instead of an `RDF::Enumerable`.

There's probably a better way to handle this than the metaprogramming approach here, but we do the same thing in at least one place in the RDF.rb core. If there are negative side effects re: method cache, we're already incurring them. I'd like to fix this up, but my feeling is that this is *Good Enough* :tm:.